core/amber/src/main/scala/edu/uci/ics/texera/web/storage/StatStorage.scala

```diff
@@ -0,0 +1,133 @@
+  val collection: MongoCollection[Document] = database.getCollection(statCollectionName)
+
+  // if the collection doesn't exist, then create it and add an index on the timestamp column
+  if (collection.countDocuments() == 0) {
+    collection.createIndex(
+      Indexes.ascending("created_at"),
+      new IndexOptions().expireAfter(timeToLive, TimeUnit.MINUTES)
+    )
+  }
+
+  TexeraWebApplication
+    .scheduleRecurringCallThroughActorSystem(
+      2.seconds,
+      statPersistFrequency.seconds
+    ) {
+      persistStats()
+    }
+
+  def persistStats(): Unit = {
+    if (AmberUtils.amberConfig.getString("storage.mode").equalsIgnoreCase("mongodb")) {
+      uncommittedStats.foreach(execution => insertOrUpdate(execution._1, execution._2))
+    }
+    uncommittedStats.clear()
+  }
+
+  def executionStatsAsDocs(
+      executionID: Int,
+      operatorStatistics: mutable.Map[String, OperatorStatistics]
+  ): Document = {
+    val doc = new Document()
+    val opsDocs = new Document()
+    operatorStatistics.foreach(op =>
+      opsDocs.append(
+        op._1,
+        new Document("operatorState", op._2.operatorState)
+          .append("aggregatedInputRowCount", op._2.aggregatedInputRowCount.toString)
+          .append("aggregatedOutputRowCount", op._2.aggregatedOutputRowCount.toString)
+      )
+    )
+    doc
+      .append("execution_ID", executionID)
+      .append("created_at", new Date())
+      .append("operatorStatistics", opsDocs)
+    doc
+  }
+
+  def insertOrUpdate(eId: Int, operator_id: String, opStats: OperatorStatistics): Unit = {
+    val workflowStats = uncommittedStats.get(eId)
+    workflowStats match {
+      // execution exists
+      case Some(operators) =>
+        operators.put(operator_id, opStats)
+        uncommittedStats.put(eId, operators)
+      // execution not yet in map
+      case None =>
+        uncommittedStats.put(eId, mutable.Map(operator_id -> opStats))
+    }
+  }
+
+  def insertOrUpdate(
+      eid: Int,
+      operatorStatistics: mutable.Map[String, OperatorStatistics]
+  ): Unit = {
+    if (executionExists(eid)) {
+      val condition: BasicDBObject = new BasicDBObject
+      condition.put(
+        "execution_ID",
+        new BasicDBObject("$eq", eid)
+      )
+      collection.replaceOne(condition, executionStatsAsDocs(eid, operatorStatistics))
+    } else {
+      collection.insertOne(executionStatsAsDocs(eid, operatorStatistics))
+    }
+  }
+
+  def executionExists(eId: Int): Boolean = {
+    val condition: Document = new Document("execution_ID", eId)
+    collection.find(condition).first() != null
+  }
+
+  def getOperatorStat(eId: Int): Map[String, OperatorStatistics] = {
+    val condition: Document = new Document("execution_ID", eId)
+    val opStatsDoc: Document = collection
+      .find(condition)
+      .first()
+
+    val opStatsMap: mutable.Map[String, OperatorStatistics] = mutable.Map()
+    val opStatsDocs: Document = opStatsDoc.get("operatorStatistics").asInstanceOf[Document]
```

Comment:
can we directly read the document as a Map?it is read as a JSON, can we maybe use `Utils.mapper` to cast it into our type?